packages/runtime/container-runtime/src/summary/summarizerHeuristics.ts

```diff
@@ -74,23 +74,55 @@ export class SummarizeHeuristicData implements ISummarizeHeuristicData {
 			summaryTime: Date.now(),
 		};
 
-		this.numSystemOpsBefore = this.numNonRuntimeOps;
-		this.numNonSystemOpsBefore = this.numRuntimeOps;
 		this.totalOpsSizeBefore = this.totalOpsSize;
 	}
 
 	public markLastAttemptAsSuccessful() {
 		this._lastSuccessfulSummary = { ...this.lastAttempt };
 
-		this.numNonRuntimeOps -= this.numSystemOpsBefore;
-		this.numSystemOpsBefore = 0;
-
-		this.numRuntimeOps -= this.numNonSystemOpsBefore;
-		this.numNonSystemOpsBefore = 0;
+		for (const queue of [this.runtimeOps, this.nonRuntimeOps]) {
+			while (
+				!queue.isEmpty() &&
+				// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+				queue.peekFront()! <= this._lastSuccessfulSummary.refSequenceNumber
+			) {
+				queue.removeFront();
+			}
+		}
 
 		this.totalOpsSize -= this.totalOpsSizeBefore;
 		this.totalOpsSizeBefore = 0;
 	}
+
+	public numOpsInSummary(referenceSequenceNumber: number): number {
+		let total = 0;
+
+		for (const queue of [this.runtimeOps, this.nonRuntimeOps]) {
+			for (const seqNum of queue.toArray()) {
+				if (seqNum <= referenceSequenceNumber) {
+					total++;
+				}
+			}
+		}
```

Comment:
Now, I'm more concerned :)  Can we take a performance hit by doing these multiple loops? Imagine a scenario where we have huge ops coming from copy and paste operations.

Reply:
Yeah, this makes me concerned too. Is it possible to do this in `recordAttempt` / `markLastAttemptAsSuccessful` where you filter the ops in the queue and set the number of ops before and after (or something like that), so that you don't have to do this calculation later on?

Reply:
@agarwal-navin @NicholasCouri I reverted this implementation as it increases the time and size complexity quite a bit for heuristics (which don't always need to be totally accurate). Yes, the container runtime now has a dependency where it calls the heuristics to record the attempt, but summarization is not broken if that call is removed as we still have the idle time and max ops mechanisms.

Reply:
yes, my vote is to keep it as it is in the latest iteration. Yes, it has the new dependency but the price to pay is small and does not change or impact anything else.  As I said during our conversation, the complexity added to keep the encapsulation is not worth it. Personally, I think it is not even a close call :)